lib/dynamoid/type_casting.rb

```diff
@@ -227,14 +227,22 @@ def process(value)
           nil
         elsif value.is_a?(String)
           dt = begin
+                 # parse first for stricter constraint failure before attempting to parse iso for precision support
+                 # e.g. DateTime.iso8601("2018-12") is valid
                  DateTime.parse(value)
+                 DateTime.iso8601(value)
```

Comment:
I don't think we need these changes in type casting. Anything that is like date/time could be casted to DateTime. Type casting isn't connected with writing/reading attributes to/from DynamoDB.

Reply:
@andrykonchin Per the comment,  the existing expected behaviour is to break for poorly parsed strings.   https://github.com/Dynamoid/dynamoid/blob/master/spec/dynamoid/type_casting_spec.rb#L143

I wrote this work around to support precision, and also keep the same erroneous expectations.  Parse is expected to throw an exception, where as constructing with iso8601 has less strict input expectations.

Reply:
I see. Well. 
```ruby
                 DateTime.parse(value)
                 DateTime.iso8601(value)
```
Here. `DateTime.parse` supports precision, doesn't it?
```ruby
irb(main):003:0> s = DateTime.now.iso8601(2)
=> "2019-07-29T22:06:24.62+03:00"
irb(main):004:0> DateTime.parse(s).usec
=> 620000
```
Here 
```ruby
            if dt_has_precision
              dt.utc
```

completely ignores `ApplicationTimeZone.utc_offset` what's wrong. If passed string without timezone specified explicitly `ApplicationTimeZone.utc_offset` should be used.

I thinks precision support could be added directly in this line:
```ruby
DateTime.new(dt.year, dt.mon, dt.mday, dt.hour, dt.min, dt.sec, offset)
```
seconds (dt.sec) could have fraction so this line could be changed in this way:
```ruby
DateTime.new(dt.year, dt.mon, dt.mday, dt.hour, dt.min, dt.se + dt.sec_fraction, offset)
```

Look at the example:
```ruby
irb(main):015:0> s = DateTime.now.iso8601(4)
=> "2019-07-29T22:24:20.8163+03:00"
irb(main):016:0> DateTime.parse(s).sec
=> 20
irb(main):018:0> DateTime.parse(s).sec_fraction
=> (8163/10000)
```